src/Migrate.php

```diff
@@ -0,0 +1,1394 @@
+    {
+        $sourceRegion = $this->sourceConnection->fetchAll(sprintf(
+            'SHOW regions LIKE %s',
+            QueryBuilder::quote($this->sourceConnection->getRegion())
+        ));
+
+        $targetRegion = $this->destinationConnection->fetchAll(sprintf(
+            'SHOW regions LIKE %s',
+            QueryBuilder::quote($this->destinationConnection->getRegion())
+        ));
+
+        $sourceAccount = sprintf(
+            '%s.%s',
+            $this->sourceConnection->getAccount(),
+            $sourceRegion[0]['region']
+        );
+        $targetAccount = sprintf(
+            '%s.%s',
+            $this->destinationConnection->getAccount(),
+            $targetRegion[0]['region']
+        );
+
+        $warehousesGrant = array_values(array_filter(
+            $mainRoleWithGrants['assignedGrants'],
+            fn($v) => $v['granted_on'] === 'WAREHOUSE' && $v['privilege'] === 'USAGE'
+        ));
+        assert(count($warehousesGrant) > 0);
+        $this->sourceConnection->grantRoleToUser($this->config->getSourceSnowflakeUser(), $mainRoleWithGrants['name']);
+        $this->destinationConnection->grantRoleToUser(
+            $this->config->getTargetSnowflakeUser(),
+            $mainRoleWithGrants['name']
+        );
+
+        foreach ($this->config->getDatabases() as $database) {
+            $schemas = $this->sourceConnection->fetchAll(sprintf(
+                'SHOW SCHEMAS IN DATABASE %s',
+                Helper::quoteIdentifier($database)
+            ));
+            foreach ($schemas as $schema) {
+                if (in_array($schema['name'], ['INFORMATION_SCHEMA', 'PUBLIC'])) {
+                    continue;
+                }
+                $this->sourceConnection->useRole($mainRoleWithGrants['name']);
+                $this->destinationConnection->useRole($mainRoleWithGrants['name']);
+                $this->sourceConnection->grantRoleToUser($this->config->getSourceSnowflakeUser(), $schema['owner']);
+                $this->destinationConnection->grantRoleToUser(
+                    $this->config->getTargetSnowflakeUser(),
+                    $schema['owner']
+                );
+                $this->sourceConnection->useRole($schema['owner']);
+
+                $tables = $this->sourceConnection->fetchAll(sprintf(
+                    'SHOW TABLES IN SCHEMA %s.%s',
+                    Helper::quoteIdentifier($database),
+                    Helper::quoteIdentifier($schema['name'])
+                ));
+                foreach ($tables as $table) {
+                    $primaryKeys = $this->sourceConnection->fetchAll(sprintf(
+                        'SHOW PRIMARY KEYS IN TABLE %s.%s.%s',
+                        Helper::quoteIdentifier($database),
+                        Helper::quoteIdentifier($schema['name']),
+                        Helper::quoteIdentifier($table['name'])
+                    ));
+                    if (!$primaryKeys) {
```

Comment:
Pokud to nemá PK, tak by se dalo aspoň zkontrolovat počet řádků z describe table, ne?

Reply:
počet řádků se kontroluje u všech tabulek 
https://github.com/keboola/project-migration-tool/blob/ondra-init/src/Migrate.php#L846

Reply:
Ha, jasně!